pkg/task/storage.go

```diff
@@ -0,0 +1,155 @@
+package task
+
+import (
+	"container/heap"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+// TaskStorage is a leveldb-backed priority queue.
+// storage is persisted before the in-memory queue to prevent inconsistency between restarts; what
+// you see in the queue is the same as what is written the database
+type TaskStorage struct {
+	Max       int
+	Path      string
+	DBOpts    *opt.Options
+	WriteOpts *opt.WriteOptions
+	ReadOpts  *opt.ReadOptions
+	tq        *TaskQueue
+	db        *leveldb.DB
+	mux       sync.Mutex
+}
+
+// Open database and load its contents into memory.
+func (s *TaskStorage) Open() error {
+	db, err := leveldb.OpenFile(s.Path, s.DBOpts)
+	if err != nil {
+		return err
+	}
+	s.db = db
+	return s.Reload()
+}
+
+// Method for closing the database. Always run Close when finished.
+func (s *TaskStorage) Close() error {
+	s.tq = nil
+	return s.db.Close()
+}
+
+// Read everything from the database into memory. Typically, you will not need to run this.
+func (s *TaskStorage) Reload() error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.tq = new(TaskQueue)
+	iter := s.db.NewIterator(nil, s.ReadOpts)
+	for iter.Next() {
+		tsk := new(Task)
+		err := json.Unmarshal(iter.Value(), tsk)
+		if err != nil {
+			return err
+		}
+		if tsk.State != TaskStateComplete {
+			heap.Push(s.tq, tsk)
+		}
+	}
+	iter.Release()
+	return iter.Error()
+}
+
+// Add an item to the priority queue
+func (s *TaskStorage) Push(tsk *Task) error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.Len() >= s.Max {
+		return fmt.Errorf("push rejected. too many items.")
+	}
+	tsk.State = TaskStateScheduled
+	key := []byte(tsk.ID)
+	val, err := json.Marshal(tsk)
+	if err != nil {
+		return err
+	}
+	err = s.db.Put(key, val, nil)
+	if err != nil {
+		return err
+	}
+	heap.Push(s.tq, tsk)
+	return nil
+}
+
+// get the next item from the priority queue
+func (s *TaskStorage) Pop() (*Task, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.Len() == 0 {
+		return nil, fmt.Errorf("attempted pop on empty queue, returning nil pointer!")
+	}
+	tsk := heap.Pop(s.tq).(*Task)
+	key := []byte(tsk.ID)
+	err := s.db.Delete(key, nil)
+	if err != nil {
+		return nil, err
```

Comment:
I think this error should either be logged only, or should make the entire system panic. But returning it like that is not good, because we did actually pop a task from our memory representation, yet failed to deliver it to the caller.